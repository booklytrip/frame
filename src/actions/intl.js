/**
 * @flow
 */

import type { Locale } from '../types';

/**
 * Change current locale to specified
 *
 * @param {String} locale - A new locale to set
 */
export const updateLocale = (locale: Locale) => ({
    type: 'INTL.UPDATE_LOCALE',
    locale,
});
