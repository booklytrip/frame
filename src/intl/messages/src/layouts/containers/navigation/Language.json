[
  {
    "id": "app.navigation.Language.en",
    "defaultMessage": "English"
  },
  {
    "id": "app.navigation.Language.lv",
    "defaultMessage": "Latvian"
  },
  {
    "id": "app.navigation.Language.ru",
    "defaultMessage": "Russian"
  }
]