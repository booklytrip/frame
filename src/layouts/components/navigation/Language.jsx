/**
 * The component represents a dropdown menu that allow user
 * to select language of the interface.
 *
 * @flow
 */

import React from 'react';
import styled from 'styled-components';
import { compose, branch, renderNothing } from 'recompose';
import { Dropdown as BaseDropdown, Icon } from 'semantic-ui-react';

import type { Locale } from '../../../types';

type LanguageProps = {
    locale: Locale,
    options: Array<Object>,
    onChange: Function,
};

const Dropdown = styled(BaseDropdown)`
    color: ${({ theme }) => theme.layout.language.color} !important;
`;

const Language = ({ locale, options, onChange }: LanguageProps) => {
    return (
        <Dropdown
            trigger={
                <span>
                    <Icon name="world" />
                    {locale.language}
                </span>
            }
            value={locale.locale}
            options={options}
            onChange={(e, { value }) => {
                onChange(value);
            }}
        />
    );
};

// Do not render the component if there is less then 2 options
export default compose(
    branch(props => props.options.length < 2, renderNothing),
)(Language);
