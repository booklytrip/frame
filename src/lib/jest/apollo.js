import React from 'react';

import { ApolloProvider as Provider } from 'react-apollo';
import ApolloClient from 'apollo-client';
const client = new ApolloClient();

export const ApolloProvider = ({ children }) => (
    <Provider client={ client }>
        { children }
    </Provider>
);
