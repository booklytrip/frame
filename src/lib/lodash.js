/**
 * @flow
 */

import _ from 'lodash';

/**
 * Return distinct values for specified key in array of objects
 *
 * @param {Array}  arr   - The list of objects
 * @param {String} field - The key name to take values for
 */
_.mixin({
    distinct: (arr: Array<any>, key: string) => (
        _.chain(arr).map(item => item[key]).uniq().sort().value()
    ),
});
