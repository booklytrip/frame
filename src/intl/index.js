import { addLocaleData } from 'react-intl';
import { reduce, mapValues } from 'lodash';

import en from 'react-intl/locale-data/en';
import lv from 'react-intl/locale-data/lv';
import ru from 'react-intl/locale-data/ru';

import * as defaultLocales from './locales';
import { locales as uikitLocales } from 'uikit';

addLocaleData([
    ...en,
    ...lv,
    ...ru,
]);

/**
 * Combine all messages from list of provides locales
 * and return single object
 */
const combineLocales = (locales) => (
    reduce(locales, (result, locale) => (
        mapValues(locale, (messages, language) => (
            { ...result[language], ...messages }
        ))
    ), {})
);

const locales = combineLocales([defaultLocales, uikitLocales]);

export { locales };
