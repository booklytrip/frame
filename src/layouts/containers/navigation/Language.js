/**
 * @flow
 */

import { compose, withProps } from 'recompose';
import { connect } from 'react-redux';
import { defineMessages, injectIntl } from 'react-intl';
import { updateLocale } from '../../../actions/intl';
import { map, find } from 'lodash';

import Language from '../../components/navigation/Language';

const intlMessages = defineMessages({
    en: {
        id: 'app.navigation.Language.en',
        defaultMessage: 'English',
    },
    lv: {
        id: 'app.navigation.Language.lv',
        defaultMessage: 'Latvian',
    },
    ru: {
        id: 'app.navigation.Language.ru',
        defaultMessage: 'Russian',
    },
});

const locales = [
    {
        locale: 'en',
        language: 'English',
        flag: 'us',
    },
    {
        locale: 'lv',
        language: 'Latvian',
        flag: 'lv',
    },
    {
        locale: 'ru',
        language: 'Russian',
        flag: 'ru',
    },
];

// const currentLocale = find(locales, { locale });

const mapStateToProps = ({ intl, project }) => ({
    currentLocale: intl.get('locale'),
    languages: project.localization.languages,
});

const props = withProps(({ currentLocale, languages, intl }) => {
    const locale = find(locales, { locale: currentLocale });
    return {
        // Provide current locale
        locale: {
            ...locale,
            language: intl.formatMessage(intlMessages[locale.locale]),
        },
        // Map list of languages defined in project settings, to list of options
        // for dropdown menu
        options: map(languages, language => {
            const locale = find(locales, { locale: language });
            return {
                flag: locale.flag,
                value: locale.locale,
                text: intl.formatMessage(intlMessages[locale.locale]),
            };
        }),
    };
});

export default compose(
    connect(mapStateToProps, { onChange: updateLocale }),
    injectIntl,
    props,
)(Language);
