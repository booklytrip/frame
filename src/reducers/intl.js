/**
 * @flow
 */

import { Map } from 'immutable';
import { saveState, loadState } from '../lib/sessionStorage';

const initialState = new Map(loadState('intl') || {});

export default function(state: Object = initialState, action: Object) {
    switch (action.type) {
        // Replace current locale with specified
        case 'INTL.UPDATE_LOCALE': {
            const newState = state.set('locale', action.locale);
            saveState('intl', newState);
            return newState;
        }
        default:
            return state;
    }
}
