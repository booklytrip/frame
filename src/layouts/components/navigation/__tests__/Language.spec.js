import React from 'react';
import { mountWithIntl, createWithIntl } from '../../../../lib/jest/intl';
import { find } from 'lodash';
import Language from '../Language';

const languages = [
    {
        language: 'en',
        text: 'English',
        flag: 'us',
    },
    {
        language: 'lv',
        text: 'Latvian',
        flag: 'lv',
    },
    {
        language: 'ru',
        text: 'Russian',
        flag: 'ru',
    },
];

describe('Language', () => {
    it('renders without crashing', () => {
        mountWithIntl(
            <Language
                locale={find(languages, { language: 'en' })}
                options={languages}
            />,
        );
    });

    it('should trigger onChange event when language is selected', () => {
        const spy = jest.fn();
        const component = mountWithIntl(
            <Language
                locale={find(languages, { language: 'en' })}
                options={languages}
                onChange={spy}
            />,
        );

        component.find('.item').first().simulate('click');
        expect(spy).toBeCalled();
    });

    describe('Snapshots', () => {
        it('should select russian language', () => {
            const rendered = createWithIntl(
                <Language
                    locale={find(languages, { language: 'ru' })}
                    options={languages}
                />,
            );
            expect(rendered).toMatchSnapshot();
        });
    });
});
