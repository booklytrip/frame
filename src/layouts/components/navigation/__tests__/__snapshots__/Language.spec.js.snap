// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Language Snapshots should select russian language 1`] = `
<div
  aria-busy={undefined}
  aria-disabled={undefined}
  aria-expanded={false}
  aria-multiselectable={undefined}
  className="ui dropdown"
  onBlur={[Function]}
  onChange={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
  role="listbox"
  tabIndex={0}
>
  <span>
    <i
      aria-hidden="true"
      className="world icon"
    />
    ru
  </span>
  <i
    aria-hidden="true"
    className="dropdown icon"
  />
  <div
    className="menu transition"
  >
    <div
      aria-checked={false}
      aria-disabled={undefined}
      aria-selected={true}
      className="selected item"
      language="en"
      onClick={[Function]}
      role="option"
      style={
        Object {
          "pointerEvents": "all",
        }
      }
    >
      <i
        className="us flag"
      />
      <span
        className="text"
      >
        English
      </span>
    </div>
    <div
      aria-checked={false}
      aria-disabled={undefined}
      aria-selected={false}
      className="item"
      language="lv"
      onClick={[Function]}
      role="option"
      style={
        Object {
          "pointerEvents": "all",
        }
      }
    >
      <i
        className="lv flag"
      />
      <span
        className="text"
      >
        Latvian
      </span>
    </div>
    <div
      aria-checked={false}
      aria-disabled={undefined}
      aria-selected={false}
      className="item"
      language="ru"
      onClick={[Function]}
      role="option"
      style={
        Object {
          "pointerEvents": "all",
        }
      }
    >
      <i
        className="ru flag"
      />
      <span
        className="text"
      >
        Russian
      </span>
    </div>
  </div>
</div>
`;
